Mathlib/Data/Real/OfDigits.lean

```diff
@@ -0,0 +1,264 @@
+    rw [Nat.floor_eq_zero]
+    exact hx.right
+  | succ n ih =>
+    rw [Finset.sum_range_succ, mul_add, pow_succ', mul_assoc, ih]
+    simp only [ofDigitsTerm, toDigits]
+    rw [show x * (b : ℝ) ^ (n + 1) = b * (b^n * x) by ring]
+    conv => rhs; rw [mul_assoc]
+    set y := (b : ℝ) ^ n * x
+    ring_nf
+    move_mul [← (b : ℝ)⁻¹]
+    have hb_zero : (b : ℝ) ≠ 0 := by
+      simpa using NeZero.ne b
+    simp only [inv_mul_cancel₀ (hb_zero), one_mul, Fin.ofNat_eq_cast, Fin.val_natCast, dvd_refl,
+      Nat.mod_mod_of_dvd, inv_pow]
+    move_mul [← ((b : ℝ)^n)⁻¹]
+    rw [inv_mul_cancel₀ (by positivity), one_mul]
+    norm_cast
+    rw [← Nat.mul_floor_div_eq_floor (y := y) (show 0 < b by omega)]
+    conv => rhs; rw [← Nat.mod_add_div ⌊(b : ℝ) * y⌋₊ b]
+
+lemma ofDigits_toDigits_partial_sum_ge {x : ℝ} {b : ℕ} [NeZero b] (hb : 1 < b)
+    (hx : x ∈ Set.Ico 0 1) {n : ℕ} :
+    x - (b⁻¹ : ℝ)^n ≤ ∑ i ∈ Finset.range n, ofDigitsTerm (toDigits x b) i := by
+  have := ofDigits_toDigits_partial_sum_eq hb hx (n := n)
+  have h_le := Nat.lt_floor_add_one (b^n * x)
+  rw [← this] at h_le
+  rw [← mul_le_mul_left (show 0 < (b : ℝ)^n by positivity),
+    mul_sub, inv_pow, mul_inv_cancel₀ (by positivity)]
+  linarith
+
+lemma ofDigits_toDigits_partial_sum_le {x : ℝ} {b : ℕ} [NeZero b] (hb : 1 < b) {n : ℕ}
+    (hx : x ∈ Set.Ico 0 1) :
+    ∑ i ∈ Finset.range n, ofDigitsTerm (toDigits x b) i ≤ x := by
+  have := ofDigits_toDigits_partial_sum_eq hb hx (n := n)
+  have h_le := Nat.floor_le (a := b^n * x) (by simp at hx; apply mul_nonneg _ hx.left; positivity)
+  rw [← this, mul_le_mul_iff_of_pos_left (by positivity)] at h_le
+  exact h_le
+
+theorem ofDigits_toDigits_HasSum (x : ℝ) (b : ℕ) [NeZero b] (hb : 1 < b) (hx : x ∈ Set.Ico 0 1) :
+    HasSum (ofDigitsTerm (toDigits x b)) x := by
+  rw [hasSum_iff_tendsto_nat_of_summable_norm]
+  swap
+  · conv => arg 1; ext i; simp; rw [abs_of_nonneg (by simp [ofDigitsTerm]; positivity)]
+    exact ofDigitsTerm_Summable
+  rw [← tendsto_sub_nhds_zero_iff]
+  apply tendsto_of_tendsto_of_tendsto_of_le_of_le (g := fun n ↦ -(b⁻¹ : ℝ)^n) (h := 0)
+  · rw [show (0 : ℝ) = -0 by simp]
+    apply Filter.Tendsto.neg
+    apply tendsto_pow_atTop_nhds_zero_of_abs_lt_one
+    rw [abs_of_nonneg (by positivity)]
+    rify at hb
+    exact inv_lt_one_of_one_lt₀ hb
+  · apply tendsto_const_nhds
+  · intro n
+    simp only [inv_pow, neg_le_sub_iff_le_add]
+    have := ofDigits_toDigits_partial_sum_ge hb hx (n := n)
+    simp only [inv_pow, tsub_le_iff_right] at this
+    linarith
+  · intro n
+    simp only [Pi.zero_apply, tsub_le_iff_right, zero_add]
+    exact ofDigits_toDigits_partial_sum_le hb hx
+
+theorem toDigits_ofDigits (b : ℕ) [NeZero b] (x : ℝ) (hb : 1 < b) (hx : x ∈ Set.Ico 0 1) :
+    ofDigits (toDigits x b) = x := by
+  simp only [ofDigits]
+  rw [← Summable.hasSum_iff]
+  · exact ofDigits_toDigits_HasSum x b hb hx
+  · exact ofDigitsTerm_Summable
+
+
+theorem ofDigits_continuous {b : ℕ} : Continuous (@ofDigits b) := by
```

Comment:
Can you directly use [continuous_tsum](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/NormedSpace/FunctionSeries.html#continuous_tsum)?